### DISCUSSION TITLE
Custom Keywords

### DISCUSSION THREAD

**Discussion by hced** [Flow-Launcher/Flow.Launcher#505] [Ideas]
(Disclaimer: Sorry if this was requested before. Tried searching for `abbreviation`, `label` and `tag` without anything.)

I'd like to be able to add an abbreviation/word for any given search result and then be able to search for that abbreviation to have them show up first in future searches. This is very useful for nudging a launcher to become more like how you want it to be. ;)

**Comment by jjw24** [comment]
Is this where you would assign abbreviation/tag to several results and then search that abbreviation/tag to have those results come up?

**Comment by hced** [comment]
It's for a single search result. Assigning your own keyword can sometimes be quite useful. E.g. `zzz` for system sleep, or similar.

**Comment by jjw24** [comment]
Hmm if you are assigning to a single result, how come you don't just search for that result? Is this for scenario where you would remember the assigned custom keyword but not the actual name of the item you are looking for?

Would this be for searching apps or files/folders ?

**Reply by Blazzycrafter** [reply]
Another example?
I use at launcher as Minecraft launcher…
Sometimes I wrote Minecraft instead of At launcher. XD

**Reply by jjw24** [reply]
This is released https://github.com/Flow-Launcher/Flow.Launcher/discussions/505#discussioncomment-5116154

**Comment by deefrawley** [comment]
If it is a search result for a single command then the Plugin Runner plugin can do this can't it?

**Comment by jjw24** [comment]
not sure, i havent tried. 

I am considering this will have another use case where you can label/tag your commonly used search results so one label/tag will bring them up and you can select from the result set or do additional search within the set

**Reply by jjw24** [reply]
i am starting to think this use case is quite useful- label/tag/add a custom keyword to your individual files, bookmark, programs. This groups them together for example my 'app design.pdf', 'github repo bookmark', 'stackoverflow page', 'vscode program' are all under a label called 'work'. so i simply search work will bring up those four results, and i can further then search within the result for github to bring up my 'work' labelled 'github repo bookmark'.

have also an option to launch the label, which launches the four results above

**Comment by hced** [comment]
Actually, my personal reason for adding different keywords than the actual search term, is just personal preference. In examples like above, I find it a much easier mnemonic to just type `zzz` for System
Sleep. I think that’s a key ingredient making a launcher excellent: the launcher should adapt to the user, rather than the user having to remember what each particular function or file was called again. You know, sometimes it just won’t stick :-) And *that’s* the real reason for why custom keywords are great. Well, they also enable you to assign much quicker letter combinations that are easier and faster to type as well. So there’s a lot to win by these.

**Comment by deefrawley** [comment]
If the target is an executable or script then Plugin Runner can do this.  To shutdown the PC (not quite sleep mode but easier to explain) you add a new Plugin Runner command, pick the c:\windows\system32\shutdown.exe file, give it the shortcut 'zzz' and save. Then in FL just type 'r' to activate the plugin and should list all your shortcuts and you can type 'zzz' or choose with the arrow keys.  If you want a shortcut to a search result set, say C:\MyPics\*.jpg then Plugin Runner won't work.

**Comment by taooceros** [comment]
I think @deefrawley 's idea is quite great. Just a few addon, you can open Files like "C:\MyPics*.jpg" with plugin runner, but by doing that you may need to use a shell to open it.
For example, the command `powershell -c & "C:\MyPics.jpg"` should be able to open the file. If you don't want a powershell window open up, you can search for args to hide powershell window, which I believe there's one. The Plugin Runner is highly Customized, which is quite an interesting plugin.

**Comment by deefrawley** [comment]
If you have Everything installed (standalone or as part of the Everything Flow plugin) then you can create Plugin Runner shortcuts to use Everything's extensive command line arguments to run a search and display the results in the Everything GUI too.

**Comment by jjw24** [comment]
> i am starting to think this use case is quite useful- label/tag/add a custom keyword to your individual files, bookmark, programs. This groups them together for example my 'app design.pdf', 'github repo bookmark', 'stackoverflow page', 'vscode program' are all under a label called 'work'. so i simply search work will bring up those four results, and i can further then search within the result for github to bring up my 'work' labelled 'github repo bookmark'.
> 
> have also an option to launch the label, which launches the four results above


Implementation wise, initial thoughts are, add the function to search with label/tag via the plugin interface, each plugin will handle the search them selves by implementing the function. Provide the actual method for this via the API instance so as long as it's implemented, the plugin will be able to use it.

Plugin via the function will save a result to a label; return a result that match the label; save the labelled result to json file.

1. Potentially can search within labelled results.

2. Fire off label combos e.g. type labels 'work + music' will show all results mentioned above plus 'music' labelled results like Spotify.

3. Something like 'open labels work + music' will trigger all matched labelled results to open at once

**Reply by taooceros** [reply]
You mean group plugins by labels?

**Reply by jjw24** [reply]
Yes both group plugins by label and @hced's original idea of labelling individual results

**Reply by taooceros** [reply]
I think it would be hard to label specific result in applition because Action (or delegate) is not serializable.

**Reply by jjw24** [reply]
What abt firing off an actual search with the saved title and return result like the query call

**Reply by pc223** [reply]
This is an interesting idea!

**Reply by taooceros** [reply]
> What abt firing off an actual search with the saved title and return result like the query call

Someone in discord is drafting a snippet plugin, which is similar to this idea.

Label for different plugins can be done by #227

**Reply by jjw24** [reply]
not sure what you mean, thought it might be better as a separate implementation to action keywords?

**Reply by taooceros** [reply]
> What abt firing off an actual search with the saved title and return result like the query call

I don't think this is quite possible to be implemented for the whole Flow (possible for a specific plugin), because it is quite possible that same query will result different results.

**Reply by jjw24** [reply]
What about firing off with exact title, not the typed query

**Reply by taooceros** [reply]
Not all plugin return result based on title search.

**Comment by jjw24** [comment]
This Custom Keywords feature request has been added. Go to Flow's settings -> Hotkeys -> Custom Query Shortcut
